src/Pate/AssumptionSet.hs

```diff
@@ -0,0 +1,340 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE OverloadedStrings   #-}
+
+
+module Pate.AssumptionSet (
+    AssumptionSet
+  , augment
+  , fromExprBindings
+  , exprBinding
+  , natBinding
+  , ptrBinding
+  , macawRegBinding
+  , fromPred
+  , toPred
+  , apply
+  , apply'
+  , isAssumedPred
+  ) where
+
+import           Control.Monad ( forM )
+import qualified Control.Monad.IO.Class as IO
+import           Data.Proxy
+
+
+import qualified Prettyprinter as PP
+import           Prettyprinter ( (<+>) )
+
+import           Data.Parameterized.Classes
+import qualified Data.Parameterized.Context as Ctx
+import qualified Data.Parameterized.Map as MapF
+
+
+import qualified What4.Interface as W4
+import qualified What4.Expr.Builder as W4B
+
+import qualified Data.Macaw.Symbolic as MS
+
+import qualified Lang.Crucible.LLVM.MemModel as CLM
+import qualified Lang.Crucible.Types as CT
+
+import           Data.Parameterized.SetF ( SetF )
+import qualified Data.Parameterized.SetF as SetF
+import qualified Pate.SimulatorRegisters as PSR
+import           Pate.Panic
+import qualified What4.ExprHelpers as WEH
+import           What4.ExprHelpers ( ExprSet, VarBindCache, ExprBindings, ppExprSet )
+import qualified Pate.ExprMappable as PEM
+
+data AssumptionSet sym =
```

Comment:
Is there some kind of invariant that this type is attempting to capture on top of `ExprSet`?

Reply:
I don't think there are any explicit invariants here. The assumption set is a collection of predicates for general predicates as well as bindings used for equality assumptions that can be applied to perform simplifications.

The state of this binding environment may need to be revisited at some point to ensure that expressions can be consistently simplified into some kind of normal form (which would imply some kind of invariant on what shape the bindings in the environment are allowed to have). At the moment the rewriter that consumes the bindings doesn't make any assumptions about their shape (i.e. it implicitly breaks rewrite loops while trying to find a rewriting fixpoint).